File: 섹션03. 2차원 배열 회전하기/03. BOJ 16935 - 배열 돌리기 3.cpp
#include <stdio.h>

#define MAX (100 + 10)

#define FLIP_UPDOWN (1)
#define FLIP_LEFTRIGHT (2)
#define CLOCKWISE (3)
#define COUNTER_CLOCKWISE (4)
#define SPLIT_CLOCKWISE (5)
#define SPLIT_COUNTER_CLOCKWISE (6)

int N, M, R;
int MAP[MAX][MAX];
int temp[MAX][MAX];

void input()
{
	scanf("%d %d %d", &N, &M, &R);

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= M; c++)
			scanf("%d", &MAP[r][c]);
}

void copyMAP(int copy[MAX][MAX], int original[MAX][MAX])
{
	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= M; c++)
			copy[r][c] = original[r][c];
}

void printMap()
{
	for (int r = 1; r <= N; r++)
	{
		for (int c = 1; c <= M; c++)
			printf("%d ", MAP[r][c]);
		putchar('\n');
	}
}

void flipUpdown()
{
	copyMAP(temp, MAP);

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= M; c++)
			MAP[r][c] = temp[N + 1 - r][c];
}

void flipLeftRight()
{
	copyMAP(temp, MAP);

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= M; c++)
			MAP[r][c] = temp[r][M + 1 - c];
}

void clockwise()
{
	copyMAP(temp, MAP);

	int t = N;
	N = M;
	M = t;

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= M; c++)
			MAP[r][c] = temp[M + 1 - c][r];
}

void counterClockwise()
{
	copyMAP(temp, MAP);

	int t = N;
	N = M;
	M = t;

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= M; c++)
			MAP[r][c] = temp[c][N + 1 - r];
}

void splitClockwise()
{
	copyMAP(temp, MAP);

	// 1, 2
	// 4, 3
	// 1 -> 2 -> 3 -> 4 -> 1
	int sr1, sc1, sr2, sc2, sr3, sc3, sr4, sc4;
	int halfN, halfM;

	halfN = (N / 2);
	halfM = (M / 2);

	sr1 = 1; sc1 = 1;
	sr2 = 1; sc2 = halfM + 1;
	sr3 = halfN + 1; sc3 = halfM + 1;
	sr4 = halfN + 1; sc4 = 1;

	for (int r = 0; r < halfN; r++)
		for (int c = 0; c < halfM; c++)
			MAP[sr2 + r][sc2 + c] = temp[sr1 + r][sc1 + c];

	for (int r = 0; r < halfN; r++)
		for (int c = 0; c < halfM; c++)
			MAP[sr3 + r][sc3 + c] = temp[sr2 + r][sc2 + c];

	for (int r = 0; r < halfN; r++)
		for (int c = 0; c < halfM; c++)
			MAP[sr4 + r][sc4 + c] = temp[sr3 + r][sc3 + c];

	for (int r = 0; r < halfN; r++)
		for (int c = 0; c < halfM; c++)
			MAP[sr1 + r][sc1 + c] = temp[sr4 + r][sc4 + c];
}

void splitCounterClockwise()
{
	copyMAP(temp, MAP);

	// 1, 2
	// 4, 3
	// 1 <- 2 <- 3 <- 4 <- 1
	int sr1, sc1, sr2, sc2, sr3, sc3, sr4, sc4;
	int halfN, halfM;

	halfN = (N / 2);
	halfM = (M / 2);

	sr1 = 1; sc1 = 1;
	sr2 = 1; sc2 = halfM + 1;
	sr3 = halfN + 1; sc3 = halfM + 1;
	sr4 = halfN + 1; sc4 = 1;

	for (int r = 0; r < halfN; r++)
		for (int c = 0; c < halfM; c++)
			MAP[sr1 + r][sc1 + c] = temp[sr2 + r][sc2 + c];

	for (int r = 0; r < halfN; r++)
		for (int c = 0; c < halfM; c++)
			MAP[sr2 + r][sc2 + c] = temp[sr3 + r][sc3 + c];

	for (int r = 0; r < halfN; r++)
		for (int c = 0; c < halfM; c++)
			MAP[sr3 + r][sc3 + c] = temp[sr4 + r][sc4 + c];

	for (int r = 0; r < halfN; r++)
		for (int c = 0; c < halfM; c++)
			MAP[sr4 + r][sc4 + c] = temp[sr1 + r][sc1 + c];
}

int main()
{
	input();

	for (int r = 0; r < R; r++)
	{
		int COMMAND;

		scanf("%d", &COMMAND);

		if (COMMAND == FLIP_UPDOWN) flipUpdown();
		else if (COMMAND == FLIP_LEFTRIGHT) flipLeftRight();
		else if (COMMAND == CLOCKWISE) clockwise();
		else if (COMMAND == COUNTER_CLOCKWISE) counterClockwise();
		else if (COMMAND == SPLIT_CLOCKWISE) splitClockwise();
		else if (COMMAND == SPLIT_COUNTER_CLOCKWISE) splitCounterClockwise();
	}

	printMap();

	return 0;
}

File: Python/섹션03. 2차원 배열 회전하기/03. BOJ 16935 - 배열 돌리기 3.py
import copy

MAX = 100 + 10

FLIP_UPDOWN = 1
FLIP_LEFTRIGHT = 2
CLOCKWISE = 3
COUNTER_CLOCKWISE = 4
SPLIT_CLOCKWISE = 5
SPLIT_COUNTER_CLOCKWISE = 6

MAP = [[0] * MAX for _ in range(MAX)]

def input_data():
    global N, M, R
    N, M, R = map(int, input().split())

    for r in range(1, N + 1):
        MAP[r][1:M + 1] = map(int, input().split())

def print_map():
    for r in range(1, N + 1):
        print(*MAP[r][1:M + 1])
                
def flip_up_down():
    temp = copy.deepcopy(MAP)
    
    for r in range(1, N + 1):
        for c in range(1, M + 1):
            MAP[r][c] = temp[N + 1 - r][c]
    
def flip_left_right():
    temp = copy.deepcopy(MAP)
    
    for r in range(1, N + 1):
        for c in range(1, M + 1):
            MAP[r][c] = temp[r][M + 1 - c]
    
def clockwise():
    global N, M
    
    temp = copy.deepcopy(MAP)
    
    N, M = M, N
    
    for r in range(1, N + 1):
        for c in range(1, M + 1):
            MAP[r][c] = temp[M + 1 - c][r]
    
def counter_clockwise():
    global N, M
    
    temp = copy.deepcopy(MAP)
    
    N, M = M, N
    
    for r in range(1, N + 1):
        for c in range(1, M + 1):
            MAP[r][c] = temp[c][N + 1 - r]
            
def split_clockwise():       
    global N, M
         
    temp = copy.deepcopy(MAP)
    
    halfN = (N // 2)
    halfM = (M // 2)
    
    sr1, sc1 = 1, 1
    sr2, sc2 = 1, halfM + 1
    sr3, sc3 = halfN + 1, halfM + 1
    sr4, sc4 = halfN + 1, 1
    
    for r in range(halfN):
        for c in range(halfM):
            MAP[sr2 + r][sc2 + c] = temp[sr1 + r][sc1 + c]
            
    for r in range(halfN):
        for c in range(halfM):
            MAP[sr3 + r][sc3 + c] = temp[sr2 + r][sc2 + c]
    
    for r in range(halfN):
        for c in range(halfM):
            MAP[sr4 + r][sc4 + c] = temp[sr3 + r][sc3 + c]
    
    for r in range(halfN):
        for c in range(halfM):
            MAP[sr1 + r][sc1 + c] = temp[sr4 + r][sc4 + c]
    
def split_counter_clockwise():    
    global N, M
            
    temp = copy.deepcopy(MAP)
    
    halfN = (N // 2)
    halfM = (M // 2)
    
    sr1, sc1 = 1, 1
    sr2, sc2 = 1, halfM + 1
    sr3, sc3 = halfN + 1, halfM + 1
    sr4, sc4 = halfN + 1, 1
    
    for r in range(halfN):
        for c in range(halfM):
            MAP[sr1 + r][sc1 + c] = temp[sr2 + r][sc2 + c]
            
    for r in range(halfN):
        for c in range(halfM):
            MAP[sr2 + r][sc2 + c] = temp[sr3 + r][sc3 + c]
    
    for r in range(halfN):
        for c in range(halfM):
            MAP[sr3 + r][sc3 + c] = temp[sr4 + r][sc4 + c]
    
    for r in range(halfN):
        for c in range(halfM):
            MAP[sr4 + r][sc4 + c] = temp[sr1 + r][sc1 + c]
            
input_data()

COMMAND = map(int, input().split())

for command in COMMAND:
    if command == FLIP_UPDOWN: flip_up_down()
    elif command == FLIP_LEFTRIGHT: flip_left_right()
    elif command == CLOCKWISE: clockwise()
    elif command == COUNTER_CLOCKWISE: counter_clockwise()
    elif command == SPLIT_CLOCKWISE: split_clockwise()
    elif command == SPLIT_COUNTER_CLOCKWISE: split_counter_clockwise()
    
print_map()
    